### PR TITLE
Feature unify ode constructor

### PR DESCRIPTION
**Pull request by jaurentz**
This pull request fixes two main issues with the ODE side of chebfun. 
 1. It allows cut off information computed in `testConvergence` to be passed to `toFunctionOut` so that the ouput of `linsolve` is an appropriately truncated `chebfun`. (See #1531.)
 2. It allows the specification of the `happinessCheck` routine that is used in `testConvergence` and in `solveivp` to be set using `cheboppref`.

### CODE REVIEW

**Comment by asgeirbirkis** [comment]
@jaurentz at first glance, this seems to bring improvements to many parts of the code, particularly when it comes to the length of solutions. Take for example the exported water droplet GUI demo. Previously, we got the following iteration info:
```
dom = [-1 1];
N = chebop(@(x,u) diff(u,2)-(u-1).*(1+diff(u).^2).^1.5, dom);
N.bc = @(x,u) [u(-1); u(1)];
options = cheboppref();
options.display = 'iter';
options.errTol = 1e-8;
options.damping = true;
options.discretization = 'collocation';
[u, info] = solvebvp(N, 0, options);

Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     3.68e-01     1.19e-01      1.0000      32        32 
 02     4.96e-02     1.35e-01      1.0000      32        32 
 03     7.23e-04     1.46e-02      1.0000      64        64 
 04     3.54e-07     4.89e-04      1.0000      64        64 
 05     7.76e-14     2.20e-07      1.0000     128       128 
--------------------------------------------------------------
```

Now we get:
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     3.68e-01     1.19e-01      1.0000      15        15 
 02     4.96e-02     1.35e-01      1.0000      27        27 
 03     7.23e-04     1.46e-02      1.0000      35        35 
 04     3.54e-07     4.89e-04      1.0000      33        35 
 05     3.55e-13     1.00e-06      1.0000      36        36 
--------------------------------------------------------------
```

However, I noticed that for the nonlinear pendulum GUI demo, something is going wrong at the final steps, and we lose out on convergence speed earlier than previously. Previously, we had

```
dom = [0 10];
N = chebop(@(x,u) diff(u,2)+sin(u), dom);
rhs = 0;
N.bc = @(x,u) [u(0)-2; u(10)-2];
x = chebfun(@(x) x, dom);
N.init =  2.*cos(2.*pi.*x./10);
options = cheboppref();
options.display = 'iter';
options.errTol = 1e-10;
options.damping = true;
options.discretization = 'collocation';
[u, infoDev] = solvebvp(N, rhs, options);
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     1.17e+00     3.20e-01      1.0000     128       128 
 02     7.09e-01     6.07e-01      1.0000     128       128 
 03     2.34e-01     3.30e-01      1.0000     128       128 
 04     7.46e-03     3.18e-02      1.0000     128       128 
 05     4.47e-05     5.99e-03      1.0000     128       128 
 06     2.06e-10     4.61e-06      1.0000      32       128 
 07     5.92e-14     2.87e-04      1.0000      64       128 
--------------------------------------------------------------
```

Now we get
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     1.17e+00     3.20e-01      1.0000      61        61 
 02     7.09e-01     6.07e-01      1.0000      81        81 
 03     2.34e-01     3.30e-01      1.0000      79        81 
 04     7.46e-03     3.18e-02      1.0000      75        81 
 05     4.47e-05     5.99e-03      1.0000      59        81 
 06     4.13e-10     9.24e-06      1.0000      23        81 
 07     2.31e-10     5.59e-01      1.0000      21        81 
 08     2.31e-10     1.00e+00      1.0000      21        81 
 09     2.31e-10     5.02e-01     4.96e-01     21        81 
 10     3.46e-10     1.69e-01     1.67e-01     25        81 
 11     5.78e-11     1.67e-01      1.0000      21        81 
--------------------------------------------------------------
```

I.e. previously, we were able to squeeze few more orders of magnitudes out before issues with machine precision set in (otherwise, iterate no. 7 should be on the order of 10^{-20}).

The following figure shows the norms of the Newton updates during the iteration
newtonconvergence

For some other demos, I actually observe an improvement, here is the pasted output for the Troesch equation.

Development:
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01   1.13e-01   3.51e-01    1.0000     128       128 
 02   1.05e-01   9.24e-01    1.0000     256       256 
 03   9.44e-02   9.02e-01    1.0000     256       256 
 04   8.15e-02   8.64e-01    1.0000     256       256 
 05   6.36e-02   7.80e-01    1.0000     256       256 
 06   3.44e-02   5.42e-01    1.0000     128       256 
 07   6.55e-03   1.90e-01    1.0000     128       256 
 08   1.61e-04   2.46e-02    1.0000     128       256 
 09   9.28e-08   5.77e-04    1.0000      32       256 
 10   7.16e-10   7.71e-03    1.0000      64       256 
```

Branch
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01   1.13e-01   3.51e-01    1.0000      87        87 
 02   1.05e-01   9.24e-01    1.0000     142       142 
 03   9.44e-02   9.02e-01    1.0000     135       142 
 04   8.15e-02   8.64e-01    1.0000     127       142 
 05   6.36e-02   7.80e-01    1.0000     118       142 
 06   3.44e-02   5.42e-01    1.0000     107       142 
 07   6.55e-03   1.90e-01    1.0000      95       142 
 08   1.61e-04   2.46e-02    1.0000      68       142 
 09   9.29e-08   5.78e-04    1.0000      22       142 
 10   4.66e-12   5.01e-05    1.0000      30       142 
```

So to summarize, I think this is clearly an improvement, but it'd be nice if we could figure out what's going on in the pendulum example before this gets merged.

**Comment by jaurentz** [comment]
@trefethen and I have been playing with this branch (actually http://github.com/chebfun/chebfun/feature-standardChop) and have noticed that `plateauCheck` often exhibits undesirable behavior. If I switch to `classicCheck` the situation improves.
```matlab
>> dom = [0 10];
>> N = chebop(@(x,u) diff(u,2)+sin(u), dom);
>> rhs = 0;
>> N.bc = @(x,u) [u(0)-2; u(10)-2];
>> x = chebfun(@(x) x, dom);
>> N.init =  2.*cos(2.*pi.*x./10);
>> options = cheboppref();
>> options.display = 'iter';
>> options.errTol = 1e-10;
>> options.damping = true;
>> options.discretization = 'collocation';
>> options.happinessCheck = @classicCheck;
>> [u, infoDev] = solvebvp(N, rhs, options);
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     1.17e+00     3.20e-01      1.0000      61        61 
 02     7.09e-01     6.07e-01      1.0000      91        91 
 03     2.34e-01     3.30e-01      1.0000      93        93 
 04     7.46e-03     3.18e-02      1.0000      91        93 
 05     4.47e-05     5.99e-03      1.0000      55        93 
 06     5.43e-10     1.21e-05      1.0000       9        93 
 07     3.94e-16     7.25e-07      1.0000       1        93 
--------------------------------------------------------------
Newton's method converged in 7 iterations.
Discretization method used: Collocation. 
Final error estimate: 3.94e-16 (differential equation) 
                      2.67e-11 (boundary conditions). 
```

I think `plateauCheck` may no longer be optimal and we are working on replacing it. Should we make a temporary checker that mimicks the current behavior in development?

**Comment by asgeirbirkis** [comment]
Hmm yes, that might be a good way for comparing, `noCutCheck` which checks whether it's happy, but always has `cutoff = length(u)`.

**Comment by asgeirbirkis** [comment]
I'm loving this new branch, further experiments I've been making are pretty suggestive that `plateauCheck` doesn't really offer much benefit over `classicCheck`, in fact, `classicCheck` seems to work better!

**Comment by jaurentz** [comment]
I'm not opposed to changing the default to `classicCheck` instead of `plateauCheck`.

**Comment by jaurentz** [comment]
It appears this is not as close as previously thought.
